### PR TITLE
Major: decoupled task queues and consumer threads, overhauled/simplif…

### PR DESCRIPTION
**Pull request by q3769**
…ied API interface

### CODE REVIEW

File: nb-configuration.xml [hunk 1]
```diff
@@ -13,6 +13,15 @@ You can copy and paste the single properties, into the pom.xml file and the IDE
 That way multiple projects can share the same settings (useful for formatting rules for example).
 Any value defined here will override the pom.xml file value but is only applicable to the current project.
 -->
-        <netbeans.hint.jdkPlatform>JDK_1.8</netbeans.hint.jdkPlatform>
+        <de-funfried-netbeans-plugins-externalcodeformatter.enableIndentationSettings>true</de-funfried-netbeans-plugins-externalcodeformatter.enableIndentationSettings>
```
Comment (uros-velickovic) [inline nb-configuration.xml:16]: should this be git ignored?

File: src/main/java/conseq4j/ConcurrentSequencer.java [hunk 2]
```diff
@@ -1,43 +1,50 @@
...
- * The above copyright notice and this permission notice shall be included in
- * all copies or substantial portions of the Software.
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
- * THE SOFTWARE.
+ * The MIT License Copyright 2021 Qingtian Wang. Permission is hereby granted, free of charge, to
```
Comment (uros-velickovic) [inline src/main/java/conseq4j/ConcurrentSequencer.java:2]: Copyright 2022

File: src/main/java/conseq4j/Conseq.java [hunk 3]
```diff
@@ -1,192 +1,289 @@
...
+import lombok.extern.java.Log;
+import org.apache.commons.pool2.ObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
 
 /**
- * @author q3769
+ * @author Qingtian Wang
  */
+@Log
```
Comment (uros-velickovic) [inline src/main/java/conseq4j/Conseq.java:37]: I think it's better to use @Slf4j annotation for logging

File: src/main/java/conseq4j/Conseq.java [hunk 4]
```diff
@@ -1,192 +1,289 @@
...
-        return this.executorCache.get(this.consistentHasher.hashToBucket(sequenceKey));
+    public <T> Future<T> submit(Object sequenceKey, Runnable task, T result) {
+        FutureHolder<T> futureHolder = new FutureHolder<>();
+        sequentialExecutors.compute(sequenceKey, (presentSequenceKey, presentExecutor) -> {
+            GlobalConcurrencyBoundedRunningTasksCountingExecutorService computed = computeExecutor(presentSequenceKey,
+                    presentExecutor);
+            futureHolder.set(computed.submit(task, result));
+            return computed;
+        });
+        return futureHolder.get();
```
Comment (uros-velickovic) [inline src/main/java/conseq4j/Conseq.java:112]: You can't return futureHolder.get() As it is async call, that value would still be null. But even if you return future, you wouldn't know when the computation is finished and future is populated.

Comment (q3769) [reply src/main/java/conseq4j/Conseq.java:112]: good point. thx!. actually, i need to make the FutureHolder to the same thing as the Future to the API client: including things like isDone() and blocking get(). no?

Comment (uros-velickovic) [reply src/main/java/conseq4j/Conseq.java:112]: correct i think those two methods are mandatory here. Or a method inside the future that can append next async job on that future. Like [Vertx ](https://vertx.io/) futures are working. But i guess then there are new, even harder problems to solve :)

**Comment by q3769** [comment]
moved to main branch

**Pull request closed without merging**